### PR TITLE
Adding Tags Support to New-AzAppServicePlan,Set-AzAppServicePlan#9895 ; adding support to Stop executing if an exception is thrown when adding a custom domain to a website.#10422;Adding support to operations for App Services not in the same resource group as the App Service Plan #10482; Apply access restriction to resources in different resource groups #10794

### LINKED ISSUE #9895: webapp:New-AzAppServicePlan - Add Tags support
## Description of the new feature
The New-AzAppServicePlan cmdlet is missing -Tag parameter. This also applies the Set-AzAppServicePlan too.

#3996 this is an old issue which was for AzureRM and was closed because apparently at that time the portal was not supporting tags on this resource type. But now the portal does support this, so should the PS cmdlet.

### PR DESCRIPTION
**Pull request by Kotasudhakarreddy**
Fixes #9895, #10422 , #10482 , #10794 

## Checklist

- [ ] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [ ] The title of the PR is clear and informative
- [ ] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [ ] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [ ] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
    - [ ] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

File: src/Websites/Websites/Cmdlets/AppServicePlans/NewAzureAppServicePlan.cs [hunk 1]
```diff
@@ -63,6 +65,9 @@ public class NewAzureAppServicePlanCmdlet : AppServicePlanBaseCmdlet
         [Parameter(Mandatory = false, HelpMessage = "Run cmdlet in the background")]
         public SwitchParameter AsJob { get; set; }
 
+        [Parameter(ParameterSetName = ParameterSet1Name, Position = 7, Mandatory = false, HelpMessage = "Tags are name/value pairs that enable you to categorize resources")]
+        public Hashtable Tags { get; set; }
```
Comment (markcowl) [inline src/Websites/Websites/Cmdlets/AppServicePlans/NewAzureAppServicePlan.cs]: Per design guidelines, his parameter should be called 'Tag'

File: src/Websites/Websites/Cmdlets/AppServicePlans/SetAzureAppServicePlan.cs [hunk 2]
```diff
@@ -49,6 +52,9 @@ public class SetAzureAppServicePlanCmdlet : AppServicePlanBaseCmdlet
         [Parameter(Mandatory = false, HelpMessage = "Run cmdlet in the background")]
         public SwitchParameter AsJob { get; set; }
 
+        [Parameter(ParameterSetName = ParameterSet1Name,Mandatory = false, HelpMessage = "Tags are name/value pairs that enable you to categorize resources")]
+        public Hashtable Tags { get; set; }
```
Comment (markcowl) [inline src/Websites/Websites/Cmdlets/AppServicePlans/SetAzureAppServicePlan.cs]: same

File: src/Websites/Websites/Utilities/WebsitesClient.cs [hunk 3]
```diff
@@ -97,11 +97,12 @@ public HostingEnvironmentProfile CreateHostingEnvironmentProfile(string resource
...
         }
 
-        public void UpdateWebApp(string resourceGroupName, string location, string webAppName, string slotName, string appServicePlan, Site siteEnvelope =null)
+        public void UpdateWebApp(string resourceGroupName, string location, string webAppName, string slotName, string appServicePlan, Site siteEnvelope =null,string appServicePlanRg=null)
         {
             var webSiteToUpdate = new Site()
             {
-                ServerFarmId = appServicePlan,
+                //Condition added to send full qualified ServerFarmId if ASP and WebApp are under different  RGs.
+                ServerFarmId = (string.IsNullOrEmpty(appServicePlanRg)&&resourceGroupName != appServicePlanRg) ? appServicePlan : siteEnvelope.ServerFarmId,
```
Comment (markcowl) [inline src/Websites/Websites/Utilities/WebsitesClient.cs]: These seem like they need some test coverage to prevent regressions

**Comment by markcowl** [comment]
@panchagnule can you review?

**Comment by panchagnula** [comment]
Add test cases for all 3 scenarios & make sure you re-record/fix the failing tests here https://dev.azure.com/azure-sdk/public/_build/results?buildId=248851&view=logs&j=5264e576-3c6f-51f6-f055-fab409685f20&t=11b6e5a1-913e-50ac-35c6-54672a9d99ea

File: src/Websites/Websites/Cmdlets/AppServicePlans/NewAzureAppServicePlan.cs (hunk 1 above)
Comment (Kotasudhakarreddy) [reply src/Websites/Websites/Cmdlets/AppServicePlans/NewAzureAppServicePlan.cs]: I have defined as same name as defined in the **Microsoft.Azure.Management.WebSites.Models.Resource** class. Please share the predefined guidelines. Will be helpful in future.

Comment (panchagnula) [reply src/Websites/Websites/Cmdlets/AppServicePlans/NewAzureAppServicePlan.cs]: @markcowl  need to clarify the usage here as a command Name vs a parameter here Tags is exposed as a parameter and not as part of the command ex. Get-AzTag? In CLI we expose this as a parameter as well & use as --tags which matches the property name. Is there some design guidelines we are missing here? Thanks!

**Comment by panchagnula** [comment]
@Kotasudhakarreddy can you do a pull-rebase to upstream master & force push may be that will re-trigger the build - somehow this is stuck.

**Comment by Kotasudhakarreddy** [comment]
> @Kotasudhakarreddy can you do a pull-rebase to upstream master & force push may be that will re-trigger the build - somehow this is stuck.

@panchagnula - done!
image

**Comment by Kotasudhakarreddy** [comment]
> Add test cases for all 3 scenarios & make sure you re-record/fix the failing tests here https://dev.azure.com/azure-sdk/public/_build/results?buildId=248851&view=logs&j=5264e576-3c6f-51f6-f055-fab409685f20&t=11b6e5a1-913e-50ac-35c6-54672a9d99ea

@markcowl - need your help in writing testcase to catch exceptions thrown by cmdlts.
Below is the test case written to handle exception scenarios and it is executing pretty well when I execute in my local. But it is failing when Test Pipeline runs during pull request validation. Based on exception, I have noticed that it is failing due to the exception thrown by cmdlt. below is FYI.
**Source of exception:**
catch (Exception e)
                {
                    **WriteWarning("Could not set custom hostname '{0}'. Details: {1}", hostName, e.ToString());**
                    return;
                }
**TestCase:**
# set Custom Host Name(s)- Failed Scenario
		$oldWebApp= Get-AzWebApp -ResourceGroupName $rgname -Name $webAppName
		$CurrentWebApp = Set-AzWebApp -ResourceGroupName $rgname -Name $webAppName -HostNames $HN
		
		#Assert
		$status
		foreach($oldHN in $oldWebApp.HostNames)
		{
			Assert-True { $CurrentWebApp.HostNames -contains $oldHN }
		}

**Comment by Kotasudhakarreddy** [comment]
> @Kotasudhakarreddy can you do a pull-rebase to upstream master & force push may be that will re-trigger the build - somehow this is stuck.
> 
> @panchagnula - done!

/azp run

File: src/Websites/Websites.Test/ScenarioTests/WebAppTests.ps1 [hunk 4]
```diff
@@ -1159,6 +1162,57 @@ function Test-SetWebApp
 		Assert-AreEqual "" $webApp.SiteConfig.PhpVersion
 		Assert-AreEqual "1.2" $webApp.SiteConfig.MinTlsVersion
 
+
```
Comment (panchagnula) [inline src/Websites/Websites.Test/ScenarioTests/WebAppTests.ps1]: GO ahead & remove this test for now & check if the build works

**Comment by panchagnula** [comment]
@markcowl the changes look good to me - the tags/ tag property is an open question - please review as well & let us know so @Kotasudhakarreddy  can make the necessary updates. Thanks!

File: src/Websites/Websites/Utilities/WebsitesClient.cs (hunk 3 above)
Comment (panchagnula) [reply src/Websites/Websites/Utilities/WebsitesClient.cs]: @Kotasudhakarreddy incase this is already done - resolve this.

File: src/Websites/Websites/Cmdlets/AppServicePlans/NewAzureAppServicePlan.cs (hunk 1 above)
Comment (markcowl) [reply src/Websites/Websites/Cmdlets/AppServicePlans/NewAzureAppServicePlan.cs]: You *must use 'Tag as the parameter name for resource tags', Tags is not allowed. This is related to PowerShell language cmdlet parameter design guidelines around plural parameter names for collection types.  It is a standard that must be followed.

**Comment by markcowl** [comment]
@Kotasudhakarreddy Please fix my original review feedback

**Comment by Kotasudhakarreddy** [comment]
Due to merge conflicts and other corrupted files closed this PR and created a new one https://github.com/Azure/azure-powershell/pull/11189 . 
cc : @panchagnula @markcowl , @ThejaChoudary

**Pull request closed without merging**